fe/fe-core/src/main/java/org/apache/doris/nereids/rules/exploration/mv/AbstractMaterializedViewRule.java

```diff
@@ -53,54 +61,80 @@ protected List<Plan> rewrite(Plan queryPlan, CascadesContext cascadesContext) {
         if (!checkPattern(queryStructInfo)) {
             return rewriteResults;
         }
 
         for (MaterializationContext materializationContext : materializationContexts) {
-            Plan mvPlan = materializationContext.getMvPlan();
-            StructInfo viewStructInfo = extractStructInfo(mvPlan, cascadesContext);
-            if (!checkPattern(viewStructInfo)) {
+            // already rewrite, bail out
+            if (queryPlan.getGroupExpression().isPresent()
+                    && materializationContext.alreadyRewrite(
+                    queryPlan.getGroupExpression().get().getOwnerGroup().getGroupId())) {
                 continue;
             }
-            if (!StructInfo.isGraphLogicalEquals(queryStructInfo.getHyperGraph(), viewStructInfo.getHyperGraph())) {
+            Plan mvPlan = materializationContext.getMtmv().getMvCache().getLogicalPlan();
+            List<StructInfo> viewStructInfos = extractStructInfo(mvPlan, cascadesContext);
+            if (viewStructInfos.size() > 1) {
+                // view struct info should only have one
+                return rewriteResults;
+            }
+            StructInfo viewStructInfo = viewStructInfos.get(0);
+            if (!checkPattern(viewStructInfo)) {
                 continue;
             }
             MatchMode matchMode = decideMatchMode(queryStructInfo.getRelations(), viewStructInfo.getRelations());
-            if (MatchMode.NOT_MATCH == matchMode) {
+            if (MatchMode.COMPLETE != matchMode) {
                 continue;
             }
             List<RelationMapping> queryToViewTableMappings =
                     RelationMapping.generate(queryStructInfo.getRelations(), viewStructInfo.getRelations());
+            if (queryToViewTableMappings == null) {
+                return rewriteResults;
+            }
             for (RelationMapping queryToViewTableMapping : queryToViewTableMappings) {
+                SlotMapping queryToViewSlotMapping = SlotMapping.generate(queryToViewTableMapping);
+                if (queryToViewSlotMapping == null) {
+                    continue;
+                }
+                LogicalCompatibilityContext compatibilityContext =
+                        LogicalCompatibilityContext.from(queryToViewTableMapping, queryToViewSlotMapping,
+                                queryStructInfo, viewStructInfo);
+                // todo outer join compatibility check
+                if (!StructInfo.isGraphLogicalEquals(queryStructInfo, viewStructInfo, compatibilityContext)) {
+                    continue;
+                }
                 SplitPredicate compensatePredicates = predicatesCompensate(queryStructInfo, viewStructInfo,
-                        queryToViewTableMapping);
+                        queryToViewSlotMapping);
                 // Can not compensate, bail out
-                if (compensatePredicates == null || compensatePredicates.isEmpty()) {
+                if (compensatePredicates.isEmpty()) {
```

Comment:
How about when predicates are exactly the same

Reply:
if the same, the compensatePredicates will be always true. call this method  `org.apache.doris.nereids.rules.exploration.mv.Predicates.SplitPredicate#isAlwaysTrue`